# CHAPA Chapter 40B Application Data 2021-2025
Author: Gia Bao (Belinda) Ngo

**Dataset:**
* TOUCHABLE CHAPA Chapter 40B Application Data 2021-2023 & 10|2023-05|2025 (https://docs.google.com/spreadsheets/d/1xcW0qveqOPD-1JvJX1jaSJjPyCVVMGbY/edit?gid=690813037#gid=690813037)
* US Cities (https://github.com/danielmiessler/SecLists/blob/master/Miscellaneous/us-cities.txt)


**Sheet used:** `23-25_data_updated`, `Geocoded Kayla 2021-2023`

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

## Load and merge 21-23 and 23-25 datasets

In [3]:
file_path = "../data/TOUCHABLE CHAPA Chapter 40B Application Data 2021-2023 & 10_2023-05_2025.xlsx"

df_2123 = pd.read_excel(file_path, sheet_name="Geocoded Kayla 2021-2023")
df_2325 = pd.read_excel(file_path, sheet_name="23-25_data_updated")

### Check columns

In [24]:
# standardize col names
df_2123.columns = df_2123.columns.str.strip().str.lower().str.replace(" ", "_").str.replace("/", "_").str.replace("?", "")
df_2325.columns = df_2325.columns.str.strip().str.lower().str.replace(" ", "_").str.replace("/", "_").str.replace("?", "")
# df_2123.columns.to_list(), df_2325.columns.to_list()

In [25]:
# check some cols in 23-25 dataset
(df_2325["application_st_address"] == df_2325["street_address"]).all(), (df_2325["application_city_town"] == df_2325["city"]).all()

(np.True_, np.True_)

final cols:
* id_number (per Applicant)
* submission_date
* submission_time
* age (age_unfilled in 21-23, age_filled in 21-23 ALL 38.5)
* disability
* chapa_races (chapa_race_ethnicity in 21-23, race_combination_normalized in 23-25)
* hispanic_latino (census_hispanic_latino in 21-23)
* census_races (census_race in 21-23)
* current_residence_town_city (current_residence_town in 21-23)
* current_residence_state
* application_source (source in 21-23, source_new in 23-25)
* hh_size
* dependents (dependents_unfilled in 21-23, dependents_filled in 21-23 ALL 0)
* hh_type (hh_type_standardized in 23-25)
* hh_income (hh_income_unfilled in 21-23, hh_income_filled in 21-23 ALL 60k)
* hh_assets (hh_assets_unfilled in 21-23, hh_assets_filled in 21-23 ALL 36k)
* income_bracket (ONLY in 21-23)
* asset_bracket (ONLY in 21-23)
* unique_id (from 23-25, per Applicant+Property)
* property_full_address (application_property in 23-25 has some age restriction info)
* property_street_address (application_st_address / street_address in 23-25)
* property_unit (property_unit_number in 21-23, application_unit in 23-25)
* property_town_city (property_town in 21-23, application_city_town / city in 23-25)
* property_state (state in BOTH)
* property_zip
* match_status
* match_type
* matched_address
* fthb_class
* msa

In [26]:
rename_col_dict_2123 = {"age_unfilled": "age",
                        "chapa_race_ethnicity": "chapa_races",
                        "census_hispanic_latino": "hispanic_latino",
                        "census_race": "census_races",
                        "current_residence_town": "current_residence_town_city",
                        "source": "application_source",
                        "dependents_unfilled": "dependents",
                        "hh_income_unfilled": "hh_income",
                        "hh_assets_unfilled": "hh_assets",
                        "property_unit_number": "property_unit",
                        "property_town": "property_town_city",
                        "state": "property_state",
                        "zip": "property_zip"}
rename_col_dict_2325 = {"application_property": "property_full_address",
                        "application_city_town": "property_town_city",
                        "application_st_address": "property_street_address",
                        "application_unit": "property_unit",
                        "source_new": "application_source",
                        "race_combination_normalized": "chapa_races",
                        "hh_type_standardized": "hh_type",
                        "state": "property_state",
                        "zip": "property_zip"}

# rename cols
df_2123 = df_2123.rename(columns=rename_col_dict_2123)
df_2325 = df_2325.rename(columns=rename_col_dict_2325)

drop_col_list_2123 = ["age_missing_flag",
                        "age_filled",
                        "dependents_missing_flag",
                        "dependents_filled",
                        "hh_income_missing_flag",
                        "hh_income_filled",
                        "hh_assets_missing_flag",
                        "hh_assets_filled"]
drop_col_list_2325 = ["submission_time",
                        "street_address",
                        "city"]

# drop cols
df_2123 = df_2123.drop(columns=drop_col_list_2123, axis=1)
df_2325 = df_2325.drop(columns=drop_col_list_2325, axis=1)

### Merge and rearrange columns

In [121]:
df = pd.concat([df_2123, df_2325], ignore_index=True).assign(current_residence_zip=np.nan)

In [122]:
# rerrange cols
col_order = ["id_number",
            "submission_date",
            "age",
            "disability",
            "chapa_races",
            "hispanic_latino",
            "census_races",
            "current_residence_town_city",
            "current_residence_state",
            "current_residence_zip",
            "application_source",
            "hh_size",
            "dependents",
            "hh_type",
            "hh_income",
            "hh_assets",
            "income_bracket",
            "asset_bracket",
            "unique_id",
            "property_full_address",
            "property_street_address",
            "property_unit",
            "property_town_city",
            "property_state",
            "property_zip",
            "match_status",
            "match_type",
            "matched_address",
            "fthb_class",
            "msa"]
df = df[col_order]

In [123]:
df.head()

,id_number,submission_date,age,disability,chapa_races,hispanic_latino,census_races,current_residence_town_city,current_residence_state,current_residence_zip,...,property_street_address,property_unit,property_town_city,property_state,property_zip,match_status,match_type,matched_address,fthb_class,msa
0,150733,2022-02-08,65,0,White/Non-Minority,0.0,White alone,Somerville,MA,NaN,...,8 Ciderpress Way,NaN,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",NaN,"Boston-Cambridge-Newton, MA-NH"
1,439181,2022-02-08,59,0,Hispanic/Latino,1.0,Unknown,Westford,MA,NaN,...,8 Ciderpress Way,NaN,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",NaN,"Boston-Cambridge-Newton, MA-NH"
2,697399,2022-02-08,NaN,0,Black or African American,0.0,Black or African American alone,Malden,MA,NaN,...,8 Ciderpress Way,NaN,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",NaN,"Boston-Cambridge-Newton, MA-NH"
3,191565,2022-02-08,NaN,0,White/Non-Minority,0.0,White alone,Haverhill,MA,NaN,...,8 Ciderpress Way,NaN,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",NaN,"Boston-Cambridge-Newton, MA-NH"
4,226436,2022-02-08,60,0,White/Non-Minority,0.0,White alone,Nashua,NH,NaN,...,8 Ciderpress Way,NaN,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",NaN,"Boston-Cambridge-Newton, MA-NH"


## Explore and clean merged dataset

In [351]:
df1 = df.copy()

In [352]:
# check dtypes and nulls
# df1.info(), df1.isnull().sum().sort_values(ascending=False)

In [353]:
print(f"total applications: {len(df1["id_number"])}, total applicants: {df1["id_number"].nunique()}")
print(f"total CHAPA properties in this dataset: {df1["property_street_address"].nunique()}")

total applications: 1739, total applicants: 1298
total CHAPA properties in this dataset: 91


In [362]:
# age: replace "u/a" with NaN and "66.5" with "66" and "4" with "0", all missing is "0", and convert to int dtype
df1["age"] = df1["age"].replace({"u/a": np.nan, 66.5: 66, 4: 0}).fillna(0).astype(int)

# disability: 1 for "Yes"; 0 for "No", "Choose not to answer", "Missing"
df1["disability"] = df1["disability"].replace({"Yes": 1, "No": 0, "no": 0, "Choose not to answer": 0, np.nan: 0}).astype(int)

# current_residence_state: standardize to "MA", change nan to "Unknown"
for x in ["Ma", "ma", "Massachusetts"]:
    df1["current_residence_state"] = df1["current_residence_state"].replace({x: "MA"})
df1["current_residence_state"] = df1["current_residence_state"].replace({np.nan: "Unknown"})

# application_source: nan to "Unknown"
df1["application_source"] = df1["application_source"].replace({np.nan: "Unknown"})

# hh_type: strip leading and trailing whitespaces
df1["hh_type"] = df1["hh_type"].str.strip().replace({np.nan: "Unknown"})

# fthb_class: 1 for "Yes"; 0 for "No" and nan
df1["fthb_class"] = df1["fthb_class"].replace({"Yes": 1, "No": 0, np.nan: 0}).astype(int)

# msa
df1["msa"] = df1["msa"].replace({"Boston-Cambridge-Newton, MA-NH": "Boston-Cambridge-Newton, MA-NH",
                                    "Boston-Cambridge-Newton, MA–NH ": "Boston-Cambridge-Newton, MA-NH",
                                    "Boston-Cambridge-Newton, MA–NH": "Boston-Cambridge-Newton, MA-NH"})

In [363]:
import re

# property_unit

# extract unit from property_full_address from 23-25 dataset
def extract_unit(addy):
    if pd.isna(addy):
        return None
    addy = addy.lower()
    match = re.search(r"unit\s*([^~]*)", addy)
    if match:
        return match.group(1).strip()
    return None

# only fill for null property_unit rows
df1.loc[df1["property_unit"].isna(), "property_unit"] = df1.loc[df1["property_unit"].isna(), "property_full_address"].apply(extract_unit)

# standardize values: remove "Unit", "#", and extra spaces
df1["property_unit"] = (
    df1["property_unit"]
    .str.replace(r"(?i)unit", "", regex=True)   # case insensitive
    .str.replace("#", "", regex=False)
    .str.strip()
)

In [364]:
# property_street_address: remove "Unit xxx" at the end of the string if exists
df1["property_street_address"] = df1["property_street_address"].str.replace(r"\s+Unit\s+\S+$", "", regex=True, case=False).str.strip()

In [365]:
# current_residence_town_city, property_town_city

# import US cities reference list
file_path = "../data/us-cities.txt"
with open(file_path) as f:
    cities = set(line.strip() for line in f if line.strip())

# find / spell check which city values are NOT in the reference list
col = "current_residence_town_city"
# col = "property_town_city"
df_cities = set(df1[col].dropna().unique())
# sorted(df_cities - cities) # cities not found in reference list

# current_residence_town_city
clean_city_dict = {"LOWELL": "Lowell",
                    "Lowell Massachusetts": "Lowell",
                    "Lynn Ma": "Lynn",
                    "Nasuha Nh": "Nashua",
                    "Arlignton": "Arlington",
                    "Marlborugh": "Marlborough",
                    "hudson": "Hudson",
                    "Tewskbury": "Tewksbury",
                    "Rosindale": "Roslindale",
                    "Miflord": "Milford",
                    "Yarmouthport": "Yarmouth Port",
                    "S. Hamilton": "South Hamilton",
                    "Burlignton": "Burlington",
                    "Newberyport": "Newburyport",
                    "North Attleboro": "North Attleborough",
                    "Walthan": "Waltham",
                    "New Bedford / Worcester": "Worcester",     # same id_number has "Worcester" for other applications
                    "New Bedford/ Worcester": "Worcester"}
df1["current_residence_town_city"] = df1["current_residence_town_city"].replace(clean_city_dict).replace({np.nan: "Unknown"})

# property_town_city
df1["property_town_city"] = df1["property_town_city"].replace({"Warehem": "Wareham"})

In [366]:
# property_zip: extract from matched_address

def extract_zip(addy):
    if pd.isna(addy):
        return None
    match = re.search(r",[^,]*?(\d{5})\s*$", addy)  # look after the last comma and extract 5 digits
    if match:
        return match.group(1)
    return None

# df1.loc[df1["property_zip"].isna(), "property_zip"] = df1.loc[df1["property_zip"].isna(), "matched_address"].apply(extract_zip)

In [367]:
# some property_full_address rows have age-restriction or first come, first serve info -> new col property_restriction

def extract_all_parentheses(text):
    if pd.isna(text):
        return None

    matches = re.findall(r"\(([^()]*)\)", text)
    if not matches:
        return None

    parts = []
    for m in matches:
        m = m.lower().strip()

        # normalize variations before splitting
        m = re.sub(r"first\s*come[,;]?\s*first\s*serve[d]?", "first come first served", m)

        # split by semicolon/comma if multiple restrictions exist
        for p in re.split(r"[;,]", m):
            clean = p.strip()
            if clean:
                parts.append(clean)

    parts = sorted(set(parts))
    return "; ".join(parts)

df1["property_restriction"] = df["property_full_address"].apply(extract_all_parentheses)

In [372]:
col = "chapa_races"
# df1[col].unique(), df1[col].isnull().sum()

In [373]:
# df1[df1["current_residence_town_city"] == "Unknown"]

In [374]:
# df1.drop(columns=["census_races"], inplace=True)

In [375]:
df1.head()

,id_number,submission_date,age,disability,chapa_races,hispanic_latino,census_races,current_residence_town_city,current_residence_state,current_residence_zip,...,property_unit,property_town_city,property_state,property_zip,match_status,match_type,matched_address,fthb_class,msa,property_restriction
0,150733,2022-02-08,65,0,White/Non-Minority,0.0,White alone,Somerville,MA,NaN,...,None,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",0,"Boston-Cambridge-Newton, MA-NH",None
1,439181,2022-02-08,59,0,Hispanic/Latino,1.0,Unknown,Westford,MA,NaN,...,None,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",0,"Boston-Cambridge-Newton, MA-NH",None
2,697399,2022-02-08,0,0,Black or African American,0.0,Black or African American alone,Malden,MA,NaN,...,None,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",0,"Boston-Cambridge-Newton, MA-NH",None
3,191565,2022-02-08,0,0,White/Non-Minority,0.0,White alone,Haverhill,MA,NaN,...,None,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",0,"Boston-Cambridge-Newton, MA-NH",None
4,226436,2022-02-08,60,0,White/Non-Minority,0.0,White alone,Nashua,NH,NaN,...,None,North Andover,MA,NaN,Match,Exact,"8 CIDERPRESS WAY, NORTH ANDOVER, MA, 01845",0,"Boston-Cambridge-Newton, MA-NH",None


### Next steps:

* hh_size: 20, 31

* clean races: cols `chapa_races`, `hispanic_latino`, `census_races`
    * fill in dummy col `hispanic_latino` from parsing through `hispanic_latino`
    * one hot encode `chapa_races`

* `hh_income`, `hh_assets`, `income_bracket`, `asset_bracket`

* geocode `matched_address`
    * some non-exact matched addresses are incorrect
    * geocode nulls

* geocode `current_residence_zip`
    * pick zipcode with most pop

* merge Prices data

* create `unique_id` for 21-23 rows
    * to connect with Resale data

## Geocode `current_residence_zip`, `matched_address`
* current_residence_zip: 1739 missing
* matched_address: 150 missing

## Create `unique_id` for 21-23 rows

## Notes for Next team

* lots of towns/neighborhoods in MA needed manual validation -> consider finding and using a txt file of MA towns and cities as a reference list